desktop/desktopentry/desktopentry.go

```diff
@@ -0,0 +1,236 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2023 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package desktopentry
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/snapcore/snapd/strutil"
+)
+
+// DesktopEntry represents a freedesktop.org desktop entry file.
+//
+// The various fields are as defined in the specification:
+// https://specifications.freedesktop.org/desktop-entry-spec/desktop-entry-spec-latest.html
+type DesktopEntry struct {
+	Filename string
+	Name     string
+	Icon     string
+	Exec     string
+
+	Hidden                bool
+	OnlyShowIn            []string
+	NotShownIn            []string
+	GnomeAutostartEnabled bool
+
+	Actions map[string]*Action
+}
+
+// Action represents an application action defined in a desktop entry file.
+type Action struct {
+	Name string
+	Icon string
+	Exec string
+}
+
+type groupState int
+
+const (
+	unknownGroup groupState = iota
+	desktopEntryGroup
+	desktopActionGroup
+)
+
+func splitList(value string) []string {
+	return strings.FieldsFunc(value, func(r rune) bool { return r == ';' })
+}
+
+// Read parses a desktop entry file.
+func Read(filename string) (*DesktopEntry, error) {
+	f, err := os.Open(filename)
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+	return parse(filename, f)
+}
+
+func parse(filename string, r io.Reader) (*DesktopEntry, error) {
+	de := &DesktopEntry{
+		Filename:              filename,
+		GnomeAutostartEnabled: true,
```

Comment:
(super nitpick) maybe a small comment here why this defaults to "true"

Reply:
This matched the behaviour of the existing `usersession/autostart` code. That in turn is based on the gnome-session code, where it will disable the desktop file if the `X-GNOME-Autostart-Enabled` key exists and is false:

https://gitlab.gnome.org/GNOME/gnome-session/-/blob/main/gnome-session/gsm-autostart-app.c#L116-129